tracing/AsyncWrap/conceptual_overview.md

```diff
@@ -0,0 +1,73 @@
+
+**registering context** is the context where a task is scheduled, for example where `nextTick` is called.
+
+**invocation context** is the context where a task is invoked.
+
+## Implementations
+
+The above concepts apply to all implementations of async\_hooks; but the implementation is different for different task scheduling mechanisms.
+
+The implementation of async\_hooks integrated in Node.js core focuses mostly on tasks and callbacks associated with UV resources. It also includes an implementation for `process.nextTick` and a JS API for use by library authors. An implementation for use with Promises which utilizes PromiseHook is also in progress.
+
+### UV Resources
+
+For UV resources (handles and requests), async\_hooks provides a mechanism to track a context ID across the lifetime of the UV resource. When the UV resource is constructed a context ID is created and "init" hooks called; when callbacks associated with the resource are executed this same context ID is passed to "before" hooks. Construction and destruction of the UV resource are taken to be adequate proxies for construction and destruction of a logical context.
+
+### JS Libraries
+
+An early form of an "Embedder API" is available in the `async\_hooks` JS module which allows a library author to create a new context (i.e. async\_id) independently of a UV resource. It is up to the library author to call `emitInit` to create the new context at the time of task registration, and to wrap the callback with `emitBefore` and `emitAfter` for execution at the time invocation. The library author can also call emitDestroy when the original context is no longer needed.
+
+Some builtin modules are instrumented with this pattern:
+
+* lib/internal/process/next\_tick.js
+* lib/timers.js
+
+#### process.nextTick
+
+process.nextTick accepts a context ID as its first parameter; this becomes the parent context ID ("trigger\_id") for `before` hooks, which are executed prior to invocation of the actual nextTick callback. The current context ID ("async\_id") for the nextTick callback itself is set when the tick callback is registered.
+
+### Promises
+
+Derives from AsyncWrap to provide context-tracking hooks for Promises. See [PR][] in review.
+
+[PR]: https://github.com/nodejs/node/pull/13000
+
+---
+
+## AsyncHook Sets
+
+```js
+const async\_hooks = require('async\_hooks');
+var myHook = async\_hooks.createHook({ init, before, after, destroy });
+myHook.enable();
+```
+
+Users can determine the context ID and take actions on the current context through hooks in hook sets. A hook set is a tuple of { init, before, after, destroy } functions which allow the user to inspect and modify the execution context at various points throughout its lifetime.
+
+  * init: called as a new context is being created. Receives new context ID (async\_id) and parent context ID (trigger_id).
+  * before: called before an async task is invoked. Receives context and parent context IDs (async\_id and trigger_id).
```

Comment:
IMO the word "task" is incorrect because it connotes that a resource will only trigger the `before()`/`after()` hooks once. I intentionally call "handles" and "requests" resources because a handle can call the `before()`/`after()` hooks zero to many times whereas a request will call them exactly once.